### DISCUSSION TITLE
Identifying HKL/UVW and Grain ID's of Active Slip Systems

### DISCUSSION THREAD

**Discussion by Savillian** [mtex-toolbox/mtex#1163] [Ask Anything] [answered]
Hi everyone,

I've been calculating Schmid factors for various loading conditions in EBSD data, and wanted to pull out the specific hkl/uvw pairs of the active slip systems for each grain.

I have a series of variables containing this information already operating correctly, but I'm overlooking how to extract the hkl/uvw and the corresponding grain ID. My output variables list the slip systems as xyz values and remain unlinked to the grain ID's as far as I can tell. I assume this is something trivial, but after a few hours of troubleshooting and diving into the MTEX website I'm at a loss. **Does anyone have any tips?**

Script included below:
```matlab
% Define Slip Systems
nb = Miller(1,1,0,EBSD('TB').CS,'hkl'); % Slip
db = Miller(1,-1,1,EBSD('TB').CS,'hkl');
nm = Miller(1,1,-2,EBSD('TB').CS,'hkl'); % Shear
dm = Miller(1,1,1,EBSD('B').CS,'hkl');
sSb = slipSystem(db,nb)
sSb = sSb.symmetrise('antipodal')
sSm = slipSystem(dm,nm)
sSm = sSm.symmetrise('antipodal')

r = normalize(vector3d(1,0,0))

% Define Stress Tensor
sigma = stressTensor.uniaxial(r)
% Rotate stress tensor into crystal reference frame
sigmaLocal = inv(Reconstructedgrains.meanOrientation) * sigma

% Calculate Schmid Factor
SFb = sSb.SchmidFactor(sigmaLocal);
SFm = sSm.SchmidFactor(sigmaLocal);

% Take the maximum along the rows
[SFMaxb,active] = max(SFb,[],2);
[SFMaxm,active] = max(SFm,[],2);

% Identify Active Slip System
sSactiveb = Reconstructedgrains.meanOrientation .* sSb(active);
sSactivem = Reconstructedgrains.meanOrientation .* sSm(active);

% Plot the maximum Schmid factor
figure
plot(Reconstructedgrains,SFMaxb,'micronbar','off','linewidth',1) % Slip
mtexColorbar location southoutside
colormap hot
saveas(gcf,[Prfx,'SchmidSlip.png'])

hold on

% Visualize the trace of the slip plane
quiver(Reconstructedgrains,sSactiveb.trace,'color','b')

% Visualize the slip direction
quiver(Reconstructedgrains,sSactiveb.b,'color','r')
hold off

saveas(gcf,[Prfx,'SchmidSlipTrace.png'])

figure
plot(Reconstructedgrains,SFMaxm,'micronbar','off','linewidth',1) % Shear
mtexColorbar location southoutside
colormap hot
saveas(gcf,[Prfx,'SchmidShear.png'])

hold on

% Visualize the trace of the slip plane
quiver(Reconstructedgrains,sSactivem.trace,'color','b')

% Visualize the slip direction
quiver(Reconstructedgrains,sSactivem.b,'color','r')
hold off

saveas(gcf,[Prfx,'SchmidShearTrace.png'])
```

**Comment by zmichels** [comment, marked as answer]
Hi,

I am probably misunderstanding... but I think I tried to do something similar before. Here's is a condensed version of the script I was running (for olivine), using the mtexdata forsterite dataset. I appended the grainset with the active slip system so that I could keep them linked.

```matlab
%% load forsterite map
mtexdata forsterite

%% grains
[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd,'angle',[10 1]*degree,'boundary','tight');
grains = grains(grains.grainSize>=3);
ebsd = ebsd(grains);
[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd,'angle',[10 1]*degree,'boundary','tight');
grains = smooth(grains,5,'moveTriplePoints');


%% phase of interest
phase = 'f';

%% forsterite grains
gf = grains(phase);


%% Slip systems
sS_AonB = slipSystem(h(1),h(2));
sS_AonC = slipSystem(h(1),h(3));
sS_ConB = slipSystem(h(3),h(2));
sS_ConA = slipSystem(h(3),h(1));
sS_all = symmetrise([sS_AonB; sS_AonC; sS_ConA; sS_ConB]);

sSLocal = gf.meanOrientation * sS_all;

%% compute Schmid factor
v = vector3d(-1,1,0);

sigma = stressTensor([0 -1 0; -1 0 0; 0 0 0]);

% visualize stress tensor
figure,
plot(sigma);


%% Schmid Factor
SF = sSLocal.SchmidFactor(sigma);

% take the maxium allong the rows
[SFMax,active] = max(SF,[],2);

% plot the maximum Schmid factor
figure,
plot(gf,SFMax,'micronbar','off','figsize','large')
cb = mtexColorbar('location', 'southoutside');


%% Active Slip system
sS_all_Active = sS_all(active);

% append the grainset
gf.prop.sS = sS_all_Active;


%% Try to classify the slip systems

% unique slip-plane normals
[Cn,iAn,iCn] = unique(gf.sS.n);

% unique Burgers vectors
[Cb,iAb,iCb] = unique(gf.sS.b);

% unique combinations
[Cs,iAs,iCs] = unique([iCn iCb],'rows');

% visualize the grains
figure,
plot(gf,iCs,'figsize','large')


%% Extract hlk/uvw matched with grain ID
% grain id
id = gf.id;
% plane normal
n = gf.sS.n;
% Burgers vector
b = gs.sS.b;
```

**Reply by Savillian** [reply]
Hi zmichels,

Thanks for the quick reply! I didn't think about appending the dataset with the slip systems, but that's a smart way to link the two together.

This process still leaves me with the slip systems as xyz coordinate pairs, and not as Miller Indices. I know there's a simple way to convert from vector3d to Miller indices (inv vector3d * crystal symmetry I believe?) but I can't get any of my processes to work out.

Thanks again!
Alec

**Reply by Savillian** [reply]
Hi again,

I managed to find a way to convert from vector3d into Miller indices.

Notionally, using the following script can complete this conversion for both slip plane and slip direction. This example only looks at slip planes.

% Define Crystal Symmetry
CSBeta = {crystalSymmetry('m-3m', [3.237 3.237 3.237], 'mineral', 'Titanium (Beta)', 'color', [0.56 0.74 0.56])};

% Calculate vector3d of active slip system plane (using prior script logic)
x = gf.sS.n;

% Calculate Miller Indices of Slip Plane
m = Miller(x,CSBeta);

However, following this process I'm presented with the error: 
"No crystal symmetry has been specified when defining a crystal direction!"

**But it mostly certainly has been specified...**

**Does anyone know where the error lies in this process?**

Thanks again!
Alec

**Reply by ralfHielscher** [reply]
You should omit the braces ```{}``` around the definition of the symmetry.

Ralf.

**Reply by Savillian** [reply]
That would do it... Must have copied that over incorrectly. Thank you Ralf!

**Reply by abhi12019** [reply]
Hi,

I have tried to identify the slip system in individual grain from your above code. But, I am not able to plot. This is my code:
```matlab
ebsd = EBSD.load(fname,CS,'interface','crc',...
  'convertEuler2SpatialReferenceFrame');
close all
figure
plot(ebsd,ebsd.bc)
colormap gray

[grains,ebsd.grainId] = calcGrains(ebsd,'angle',15*degree);
% remove small grains
ebsd(grains(grains.diameter<=5)) = [];
% redo grain reconstruction
[grains,ebsd.grainId] = calcGrains(ebsd,'angle',15*degree);
%calculating from basics
sS=slipSystem.fcc(ebsd.CS);
sS=sS.symmetrise;

%rotate slip systems into specimen coordinates
sSlocal=grains.meanOrientation*sS;

%compute Schmidt factor
sigma=stressTensor.uniaxial(vector3d.X);
SF=sSlocal.SchmidFactor(sigma);

%take the maximum along the rows

[SFMax, active]=max(SF,[],2);

%plot the maximum schmidt factor
figure
plot(grains, SFMax, 'micronbar', 'off','linewidth',2,'coordinates','on')
mtexColorbar location southoutside


%take the active slip system and rotate it in specimen coordinates
sSactive=grains.meanOrientation.*sS(active);
hold on
%visualize the trace of the slip plane
quiver(grains, sSactive.trace,'color','b')

% and the slip direction
quiver(grains,sSactive.b,'color','r')
hold on
text(grains,int2str(grains.id))
hold off

%% Active Slip system
sS_all_Active = sS(active);

% append the grainset
gf.prop.sS = sS_all_Active


%% Try to classify the slip systems

% unique slip-plane normals
[Cn,iAn,iCn] = unique(gf.prop.sS.n);

% unique Burgers vectors
[Cb,iAb,iCb] = unique(gf.prop.sS.b);

% unique combinations
[Cs,iAs,iCs] = unique([iCn iCb],'rows');

% visualize the grains
figure
plot(gf,iCs,'figsize','large')
```
I am seeing the error in the last line. I don't know why this is coming.  The error is 
```
Error using plot
Invalid data argument.

Error in EBSD_confocal_specimen_1_site2_mapdat2_corrected (line 95)
plot(gf,iCs,'figsize','large')
```
Thank you in advance for your help.

Thanks again,
Abhishek

**Reply by ralfHielscher** [reply]
In your code `gf` is nowhere defined. I have no idea what it should be.
Ralf

**Reply by abhi12019** [reply]
Hi Ralf,

Thank you for pointing it out. I have changed it now. But, I am not getting the slip system information on the grains map. I am getting this plot by running the above code.
![image](https://github.com/mtex-toolbox/mtex/assets/82941844/5aab4e68-48d2-4337-8eb7-c238cf96a7d1)

How do I plot the slip system information on individual grain, so that I now this is the slip system is activated in this grain. Thank you for your help.

Thanks again,
Abhishek

**Reply by Savillian** [reply]
What specific slip system information are you looking for? The specific plane-pair permutation of the FCC system?

**Reply by abhi12019** [reply]
Hi Savillian,

Yes I want to plot the slip system of FCC (activated) on the grains map. I can see the plan normal and direction on the grain, but I am not seeing the slip system information on the each grain. This is my plot coming:
![image](https://github.com/mtex-toolbox/mtex/assets/82941844/082b30c3-6293-4323-9c60-bbc21964e0c3)

I want to see the activated slip system on the grains map. Thank you for your help

Thanks,
Abhishek

**Reply by Savillian** [reply]
Looking back at my initial script for this discussion topic, it should be possible to pull out the slip system with the maximum Schmid factor.

I think this value might be stored in your SFMax variable or one of the related calculations. 

In case the issue lies with the activated slip systems hiding in another variable we can't identify, you might be better off defining the slip systems explicitly like I did for the HCP system. 

This might still take some clicking through variables in the workspace to pull out, but once you find the workspace location, extracting the info should be simple (e.g., could be something like SFMax.active.h for the first value of the active slip system). Your code is already doing it, just have to find where that is.

I'd also watch out for any issues in bringing the slip systems into the specimen symmetry orientation. If this doesn't compute correctly, symmetrically equivalent slip systems in different grains might have different hkl values than they otherwise should (e.g., the same direction of slip in one grain is the (111) [-110] while in another it might be the (111) [01-1] because the unit cell reference frame is rotated relative to one another).
